pkg/streamingpromql/engine_test.go

```diff
@@ -159,3 +157,88 @@ func TestOurTestCases(t *testing.T) {
 		})
 	}
 }
+
+// Testing instant queries that return a range vector is not supported by Prometheus' PromQL testing framework,
+// and adding support for this would be quite involved.
+//
+// So instead, we test these few cases here instead.
+func TestRangeVectorSelectors(t *testing.T) {
+	opts := NewTestEngineOpts()
+	streamingEngine, err := NewEngine(opts)
+	require.NoError(t, err)
+
+	prometheusEngine := promql.NewEngine(opts)
+
+	baseT := timestamp.Time(0)
+	storage := promql.LoadedStorage(t, `
+		load 1m
+			some_metric{env="1"} 0+1x4
+			some_metric{env="2"} 0+2x4
+	`)
+	t.Cleanup(func() { require.NoError(t, storage.Close()) })
+
+	testCases := map[string]struct {
+		expr     string
+		expected *promql.Result
+		ts       time.Time
+	}{
+		"matches series with points in range": {
+			expr: "some_metric[1m]",
+			ts:   baseT.Add(2 * time.Minute),
+			expected: &promql.Result{
+				Value: promql.Matrix{
+					{
+						Metric: labels.FromStrings("__name__", "some_metric", "env", "1"),
+						Floats: []promql.FPoint{
+							{T: timestamp.FromTime(baseT.Add(time.Minute)), F: 1},
+							{T: timestamp.FromTime(baseT.Add(2 * time.Minute)), F: 2},
+						},
+					},
+					{
+						Metric: labels.FromStrings("__name__", "some_metric", "env", "2"),
+						Floats: []promql.FPoint{
+							{T: timestamp.FromTime(baseT.Add(time.Minute)), F: 2},
+							{T: timestamp.FromTime(baseT.Add(2 * time.Minute)), F: 4},
+						},
+					},
+				},
+			},
+		},
+		"matches no series": {
+			expr: "some_nonexistent_metric[1m]",
+			ts:   baseT,
+			expected: &promql.Result{
+				Value: promql.Matrix{},
+			},
+		},
+		"no samples in range": {
+			expr: "some_nonexistent_metric[1m]",
```

Comment:
Have you looked into this too?

Reply:
Yes, sorry, overlooked this: from the perspective of the engine, a metric with no samples at all in the TSDB, and a metric with no samples in the time range are the same, as the TSDB just won't return the series in both cases.

Or is there an edge case I'm missing here?

Reply:
I don't think you're missing anything. I thought that the test case description would better match the actual test case if we query an existing metric with no samples for the requested time range. IIRC, in Mimir we return full chunks from ingesters and store-gateway (without cutting them to query time boundaries), so we expect PromQL engine to filter out samples outside the queried time range. I was just wondering to mimick the same here, but don't spend too much time on this if it's not as trivial as querying `some_metric` instead of `some_nonexistent_metric`.

Reply:
Makes sense, I've changed this to use `some_metric` in 71298d7.